Sources/Document/Change/ChangeID.swift

```diff
@@ -0,0 +1,113 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+/**
+ * `ChangeID` is for identifying the Change. This is immutable.
+ */
+struct ChangeID {
+    /**
+     * `initial` represents the initial state ID. Usually this is used to
+     * represent a state where nothing has been edited.
+     */
+    static let initial = ChangeID(clientSeq: 0, lamport: 0, actor: ActorIDs.initial)
+
+    // `serverSeq` is optional and only present for changes stored on the server.
+    private var serverSeq: Int64?
+
+    private let clientSeq: Int
+    private var lamport: Int64
+    private var actor: ActorID?
+
+    init(clientSeq: Int, lamport: Int64, actor: ActorID? = nil) {
+        self.clientSeq = clientSeq
+        self.lamport = lamport
+        self.actor = actor
+    }
+
+    /**
+     * `next` creates a next ID of this ID.
+     */
+    func next() -> ChangeID {
+        return ChangeID(clientSeq: self.clientSeq + 1, lamport: self.lamport + 1, actor: self.actor)
+    }
+
+    /**
+     * `syncLamport` syncs lamport timestamp with the given ID.
+     *
+     * {@link https://en.wikipedia.org/wiki/Lamport_timestamps#Algorithm}
+     */
+    mutating func syncLamport(with otherLamport: Int64) {
+        let lamport = otherLamport > self.lamport ? otherLamport : self.lamport + 1
+        self.lamport = lamport
+    }
+
+    /**
+     * `createTimeTicket` creates a ticket of the given delimiter.
+     */
+    func createTimeTicket(delimiter: UInt32) -> TimeTicket {
+        return TimeTicket(lamport: self.lamport, delimiter: delimiter, actorID: self.actor)
+    }
+
+    /**
+     * `setActor` sets the given actor.
+     */
+    mutating func setActor(_ actorID: ActorID) {
+        self.actor = actorID
+    }
+
+    /**
+     * `getClientSeq` returns the client sequence of this ID.
+     */
+    func getClientSeq() -> Int {
+        return self.clientSeq
+    }
+
+    /**
+     * `getLamport` returns the lamport clock of this ID.
+     */
+    func getLamport() -> Int64 {
+        return self.lamport
+    }
+
+    /**
+     * `getLamportAsString` returns the lamport clock of this ID as a string.
+     */
+    func getLamportAsString() -> String {
+        return "\(self.lamport)"
+    }
+
+    /**
+     * `getActorID` returns the actor of this ID.
+     */
+    func getActorID() -> String? {
+        return self.actor
+    }
+
+    /**
+     * `getStructureAsString` returns a string containing the meta data of this ID.
+     */
+    func getStructureAsString() -> String {
+        var actor: String
+        if let value = self.actor {
+            actor = String(value.substring(from: 22, to: 24))
```

Comment:
@hackerwins What are the 22 and 24? I want to convert the numbers to constants.

Reply:
`ActorID` is an ObjectID generated by MongoDB, which is long and difficult to check, so we use only the last two characters while debugging.

```
ObjectId("507f1f77bcf86cd799439011") -> "11"
```

Reply:
Or We can extend the ActorID

Reply:
> `ActorID` is an ObjectID generated by MongoDB, which is long and difficult to check, so we use only the last two characters while debugging.
> 
> ```
> ObjectId("507f1f77bcf86cd799439011") -> "11"
> ```

OK, I changed it to show the last two characters.